ValveResourceFormat/KeyValues/KeyValues3.cs

```diff
@@ -0,0 +1,526 @@
+                parser.objStack.Peek().AddProperty(parser.currentName, new KVValue(KVType.FLAGGED_STRING, parser.currentString.ToString()));
+                return;
+            }
+
+            parser.currentString.Append(c);
+        }
+
+        //Read comments
+        private static void ReadComment(char c, Parser parser)
+        {
+            //Check for multiline comments
+            if (parser.currentString.Length == 1 && c == '*')
+            {
+                parser.stateStack.Pop();
+                parser.stateStack.Push(State.COMMENT_BLOCK);
+            }
+
+            //Check for the end of a comment
+            if (c == '\n')
+            {
+                parser.stateStack.Pop();
+                return;
+            }
+
+            if (c != '\r')
+            {
+                parser.currentString.Append(c);
+            }
+        }
+
+        //Read a comment block
+        private static void ReadCommentBlock(char c, Parser parser)
+        {
+            //Look for the end of the comment block
+            if (c == '/' && parser.currentString.ToString().Last() == '*')
+            {
+                parser.stateStack.Pop();
+            }
+
+            parser.currentString.Append(c);
+        }
+
+        //Get the next char from the filestream
+        private static char NextChar(Parser parser)
+        {
+            //Check if there are characters in the buffer, otherwise read a new one
+            if (parser.charBuffer.Count > 0)
+            {
+                return parser.charBuffer.Dequeue();
+            }
+            else
+            {
+                return (char)parser.fileStream.Read();
+            }
+        }
+
+        //Skip the next X characters in the filestream
+        private static void SkipChars(Parser parser, int num)
+        {
+            for (int i = 0; i < num; i++)
+            {
+                NextChar(parser);
+            }
+        }
+
+        //Utility function
+        private static string PeekString(Parser parser, int length)
+        {
+            char[] buffer = new char[length];            
+            for (int i = 0; i < length; i++)
+            {
+                if (i < parser.charBuffer.Count)
+                {
+                    buffer[i] = parser.charBuffer.ElementAt(i);
+                }
+                else
+                {
+                    buffer[i] = (char)parser.fileStream.Read();
+                    parser.charBuffer.Enqueue(buffer[i]);
+                }
+            }
+            return String.Join("", buffer);
+        }
+    }
+
+    //Data structure to contain the file contents
+    public class KVFile
+    {
+        public String Header;
+        public KVObject Root;
+
+        public KVFile(String header, KVObject root)
+        {
+            this.Header = header;
+            this.Root = root;
+        }
+
+        //Serialize the contents of the data structure
+        public string Serialize()
```

Comment:
Does this need to be it's own function, or would it be more appropriate to override `ToString`?